content/blog/angular-extend-class/index.md

```diff
@@ -0,0 +1,842 @@
+})
+class BaseComponent implements OnInit, OnDestroy {
+  // ...
+}
+```
+
+That solves one problem, but still leaves one present with using `@Component`: you must add a declaration of the `BaseComponent` into an `NgModule`. Otherwise, you'll end up with the following error during compilation:
+
+```
+BaseComponent is not declared in any Angular module 
+```
+
+To solve this, we can either import `BaseComponent` in an `NgModule` or, alternatively, mark `BaseComponent` as an abstract class:
+
+```typescript
+@Component({
+  template: ''
+})
+abstract class BaseComponent implements OnInit, OnDestroy {
+  // ...
+}
+```
+
+## `@Injectable` is an alternative of an abstract class {#injectable}
+
+[Since Angular 10 you can now use `@Injectable` to declare your `BaseComponent` instead](https://angular.io/guide/migration-injectable). This sidesteps the problem of having to mark a component `class` as abstract because even without it `Injectable`s do not need to be declared in a module:
+
+```typescript
+@Injectable()
+class BaseComponent implements OnInit, OnDestroy {
+  // ...
+}
+```
+
+You might expect there to be some migration of `AppComponent` when you're using `@Injectable` instead of `@Component` for the `BaseComponent`, but alas there is not.
+
+```typescript
+@Injectable()
+class BaseComponent implements OnInit {
+  ngOnInit() {
+    console.log('I AM BASE COMPONENT');
+  }
+}
+
+@Component({
+  selector: 'app-root',
+  template: `
+    <p>Test</p>
+  `,
+})
+class AppComponent extends BaseComponent {
+}
+```
+
+That said, [while using `@Injectable` is explicitly supported](https://github.com/angular/angular/issues/41229#issuecomment-800310757), it's very hacky to use this in place of `@Component`. This is because Angular's `@Injectable`s do not support lifecycle methods without a `Component` that extends it.
+
+As such, we'll be sticking to the `abstract` class solution.
+
+# Overwriting Lifecycle Methods {#lifecycle-methods}
+
+If you recall from our quick overview of what a base class does, you can replace the base class implementation of both methods and properties.
+
+The same is true for lifecycle methods, since they're just a type of method on the component class instance.
+
+```typescript
+@Component({
+  template: ''
+})
+abstract class BaseComponent implements OnInit {
+  ngOnInit() {
+    console.log('I AM BASE COMPONENT');
+  }
+}
+
+@Component({
+  selector: 'app-root',
+  template: `
+    <p>Test</p>
+  `,
+})
+class AppComponent extends BaseComponent implements OnInit {
+  override ngOnInit() {
+    console.log("And I am the AppComponent")
+  }
+}
+```
+
+The downside here, however, is that `ngOnInit` on `AppComponent` will no longer call the `BaseComponent`'s `ngOnInit` logic. After all, what if you wanted to simply _add_ behavior to `ngOnInit`, rather than replace it entirely?
+
+Luckily, we can use the `super` keyword to refer to the base class instance and call the original method inside of the overwritten method:
```

Comment:
praise: Good tip!

Reply:
suggestion: Consider mentioning that if we forget to add `override`, TypeScript will tell us that we're attempting to override an existing method on a base class and thus we must mark it with `override`. This in turn tells us that we should probably add a `super` call.